### PR TITLE
Unifying the FX and TS Frontends

### PR DESCRIPTION
**Pull request by narendasan**
# Description

This PR will merged the backends for FX and TS to use a common set of resources including sharing the Torch-TRT runtime which will enable users to post FX compilation, TS and serialize compiled modules and add C++ execution to FX.

Fixes #1372 

## Type of change

Please delete options that are not relevant and/or add your own.

- New feature (non-breaking change which adds functionality)

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/fx/trt_module_next.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+from typing import Any, List, Sequence
+
+import torch
+
+from torch.classes.tensorrt import Engine
+from torch.ops.tensorrt import execute_engine
+
+from torch_tensorrt import (_C, Device)
+
+class TRTModule(torch.nn.module):
```
Comment (yinghai) [inline py/torch_tensorrt/fx/trt_module_next.py:11]: Is this some staging module?

Comment (narendasan) [reply py/torch_tensorrt/fx/trt_module_next.py:11]: Yeah this would be an alternate to the existing TRTModule so it is opt-in for users for the next release for evaluation.

**Comment by narendasan** [comment]
@frank-wei can you take a look at this PR? Hopefully it doesnt introduce any breaking changes

**Comment by narendasan** [comment]
Also @peri044 can you review on the TS side?

**Review by frank-wei — COMMENTED** [review]
Generally, the PR looks good to me. It provides a way for users who wants to deploy in C++. But user can also choose to use fx TRTModule to wrap the engine and jit trace later.

File: py/torch_tensorrt/_TRTModule.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
+from operator import truediv
+from typing import Any, List, Sequence, Tuple
+
+import torch
+from torch_tensorrt import _C
+from torch_tensorrt._Device import Device
+
+
+class TRTModule(torch.nn.Module):
```
[the remark this answers is not in the document]
Comment (narendasan) [reply py/torch_tensorrt/_TRTModule.py:9]: How about `TRTModuleNext`?

[the remark this answers is not in the document]
Comment (frank-wei) [reply py/torch_tensorrt/_TRTModule.py:9]: Good for me.

**Review by frank-wei — APPROVED** [review]

File: py/torch_tensorrt/_TRTModuleNext.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+from typing import Any, List, Sequence, Tuple
+
+import torch
+from torch_tensorrt import _C
+from torch_tensorrt._Device import Device
+
+logger = logging.getLogger(__name__)
+
+
+class TRTModuleNext(torch.nn.Module):
```
Comment (gs-olive) [inline py/torch_tensorrt/_TRTModuleNext.py:12]: Consider adding a `TRTModuleNext.close()` function, as performing inference with the module and then closing the python session results in an error:
```python
ERROR: [Torch-TensorRT - Debug Build] - 3: [engine.cpp::~Engine::305]
Error Code 3: API Usage Error (Parameter check failed at: runtime/api/engine.cpp::~Engine::305,
condition: objectCount() == 0. Destroying an engine object before destroying objects it created
leads to undefined behavior.)
```

**Review by gs-olive — COMMENTED** [review]
Added a few minor comments, otherwise looks good and is great to use! Adding this review as comments since these are primarily usability-based suggestions.

File: py/torch_tensorrt/ts/_compiler.py [hunk 4]
```diff
@@ -156,7 +156,7 @@ def convert_method_to_trt_engine(
     dla_global_dram_size=536870912,
     truncate_long_and_double=False,
     calibrator=None,
-) -> str:
+) -> bytearray:
```
Comment (gs-olive) [inline py/torch_tensorrt/ts/_compiler.py:159]: Could change the `method_name` argument in this function to default to `forward`

File: py/torch_tensorrt/_TRTModuleNext.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        input_binding_names (List[str]): List of input TensorRT engine binding names in the order they would be passed to the TRT modules
+        output_binding_names (List[str]): List of output TensorRT engine binding names in the order they should be returned
+    """
+
+    def __init__(
+        self,
+        name: str = "",
+        serialized_engine: bytearray = bytearray(),
+        input_binding_names: List[str] = [],
+        output_binding_names: List[str] = [],
```
[the remark this answers is not in the document]
Comment (narendasan) [reply py/torch_tensorrt/_TRTModuleNext.py:37]: Yes there is an assumed convention, but no solid way to know every time since TensorRT does not guarantee binding order. Added docs to address this.
